messaging/kafka-streams-reactive-messaging/src/main/java/io/quarkus/ts/messaging/kafka/reactive/streams/shutdown/ReadinessObserver.java

```diff
@@ -0,0 +1,39 @@
+package io.quarkus.ts.messaging.kafka.reactive.streams.shutdown;
+
+import jakarta.enterprise.event.Observes;
+
+import io.quarkus.logging.Log;
+import io.quarkus.runtime.StartupEvent;
+import io.smallrye.reactive.messaging.health.HealthReport;
+import io.smallrye.reactive.messaging.providers.extension.HealthCenter;
+
+public class ReadinessObserver {
+
+    private static final int SLEEP_PERIOD = 250;
+
+    void observe(@Observes StartupEvent ignored, HealthCenter healthCenter) throws InterruptedException {
+        // Quarkus doesn't wait till channels are ready, but we don't want to start working till we can use Kafka
+        // see https://github.com/quarkusio/quarkus/issues/41441 for more info
+        Log.info("Delaying application startup until incoming channel 'slow' and outgoing channel 'slow-topic' are ready");
+        while (true) {
+            final HealthReport healthReport = healthCenter.getReadiness();
+            if (isChannelReady(healthReport, "slow-topic")) {
+                if (isChannelReady(healthReport, "slow")) {
+                    Log.info("Channels 'slow' and 'slow-topic' are ready, proceeding with application startup");
+                    return;
+                } else {
+                    Log.info("Channel 'slow' is not ready");
+                }
+            } else {
+                Log.info("Channel 'slow-topic' is not ready");
+            }
+            Log.info("Going to sleep for " + SLEEP_PERIOD + " milliseconds");
+            Thread.sleep(SLEEP_PERIOD);
+        }
+    }
+
+    private static boolean isChannelReady(HealthReport healthReport, String channelName) {
+        return healthReport.getChannels().stream()
```

Comment:
How slow/fragile is `getChannels` method? Checking both channels on the same call may be a better idea. And it will remove nested `if`s from the `observe` method

Reply:
it's not fragile at all (well, it is intercepted with some Quarkus magic, but that is not relevant for us). I had it separate in order to have better logging. I will merge it into one, it is not important logging message.

Reply:
Done.